### ISSUE TITLE
[Bug] android-app:// referer (Google Discover/Android apps) silently dropped by sanitize_url() since v5.4.0

### ISSUE THREAD

**Issue by parhumm** [slimstat/wp-slimstat#306]
## Summary

`android-app://` referers (Google Discover, Google app, Twitter app, Reddit app, etc.) are silently dropped in v5.4.0+ — the `referer` column is stored empty and the visit is bucketed as "Direct Traffic". Worked correctly in v5.3.5.

Reported via wp.org support forum (reporter: dxylott54). Confirmed regression introduced by the v5.4.0 tracker refactor.

## Root Cause

**Fault location**: `src/Tracker/Ajax.php:129` — `$stat['referer'] = sanitize_url($referer);`

WordPress's `sanitize_url()` → `esc_url_raw()` → `esc_url($url, null, 'db')` returns `''` for any URL whose scheme is not in `wp_allowed_protocols()`. The default allowlist is `http, https, ftp, ftps, mailto, news, irc, irc6, ircs, gopher, nntp, feed, telnet, mms, rtsp, sms, svn, tel, fax, xmpp, webcal, urn`. **`android-app` is not on it**, so `android-app://com.google.android.googlequicksearchbox/` becomes `''`.

The same call in the server-side fallback path (`src/Tracker/Processor.php:188`, `sanitize_url(wp_unslash($_SERVER['HTTP_REFERER']))`) strips it identically, so neither JS-tracked nor server-tracked Discover hits survive.

Ironically, `src/Tracker/Processor.php:199` explicitly allow-lists `android-app` in its post-storage XSS scheme check:

```php
if (isset($parsed_url['scheme']) && ('' !== $parsed_url['scheme'] && '0' !== $parsed_url['scheme']) && !in_array(strtolower($parsed_url['scheme']), ['http', 'https', 'android-app'])) {
    $stat['notes'][] = sprintf(__('Attempted XSS Injection: %s', 'wp-slimstat'), $stat['referer']);
    unset($stat['referer']);
}
```

So the design intent was clearly to accept `android-app://` — the bug is the upstream `sanitize_url()` stripping the value before this check ever runs.

**Causal chain**:
Android Google app sets `Referer: android-app://com.google.android.googlequicksearchbox/` → JS tracker reads `document.referrer`, base64URL-encodes, posts as `data_js['ref']` → `Ajax.php:105` decodes → `Ajax.php:106` `parse_url` returns `[scheme=android-app, host=com.google.android.googlequicksearchbox]` (valid) → `Ajax.php:118` host regex matches (alphanumeric + dots, all parts ≤63 chars) → **`Ajax.php:129` `sanitize_url($referer)` returns `''`** → `$stat['referer']` is empty → fallback at `Processor.php:187-188` also applies `sanitize_url()` → same empty result → DB row stored with empty `referer`.

**5 Whys**:
1. Why no Discover referer? → `$stat['referer']` is empty when stored.
2. Why empty? → `sanitize_url('android-app://...')` returns `''`.
3. Why does sanitize_url strip it? → `android-app` is not in WordPress's default `wp_allowed_protocols()`.
4. Why is `sanitize_url()` applied to JS-decoded referer in 5.4.x but not 5.3.5? → Added during the v5.4.0 tracker refactor (commit `d6b8bb9c`), uncoordinated with the `android-app` allow-list that already exists in Processor.
5. Why wasn't this caught? → No E2E/integration test for non-`http(s)` referer schemes; Processor.php's explicit `android-app` allow-list gave the false impression the path was covered.

## Reproduction

**Preconditions**: WordPress 5.6+, wp-slimstat 5.4.0–5.4.12, default tracking mode (JS or server-side). No filter has added `android-app` to `wp_allowed_protocols`.

1. From an Android device, open the Google app (Discover feed) and tap a card linking to a wp-slimstat-tracked site.
2. Open Slimstat → Right Now / Audience → Top Referrers / Access Log.

**Expected** (v5.3.5): visit row shows referer `android-app://com.google.android.googlequicksearchbox/`.
**Actual** (v5.4.0+): `referer` column is empty; the visit is counted as Direct Traffic, no Discover attribution.

### Minimal PHP repro

```php
// $referer is what would come out of Utils::base64UrlDecode($data_js['ref'])
$referer = 'android-app://com.google.android.googlequicksearchbox/';

$parsed = parse_url($referer);
// $parsed === ['scheme' => 'android-app', 'host' => 'com.google.android.googlequicksearchbox', 'path' => '/']

$host_re = '/^[a-zA-Z0-9]([a-zA-Z0-9\-]{0,61}[a-zA-Z0-9])?(\.[a-zA-Z0-9]([a-zA-Z0-9\-]{0,61}[a-zA-Z0-9])?)*$/';
preg_match($host_re, $parsed['host']); // 1 — passes Ajax.php:118 host check

$stored = sanitize_url($referer);
// $stored === ''   <-- bug: scheme not in wp_allowed_protocols(), value stripped
```

## Code References

| File | Lines | Description |
|------|-------|-------------|
| `src/Tracker/Ajax.php` | 129 | `sanitize_url($referer)` strips `android-app://` before storage (root cause) |
| `src/Tracker/Ajax.php` | 102-130 | Full referer validation block; the `parse_url` + host-regex + length guards are sound, but the final `sanitize_url` undoes them for non-`http(s)` schemes |
| `src/Tracker/Processor.php` | 187-189 | Server-side `$_SERVER['HTTP_REFERER']` fallback applies the same `sanitize_url(wp_unslash(...))` — strips `android-app://` too |
| `src/Tracker/Processor.php` | 199 | Post-storage XSS check **already allow-lists** `android-app` — design intent is clear |

## Suggested Fix Directions

1. **Minimum (surgical)** — replace the `sanitize_url()` call at `Ajax.php:129` and `Processor.php:188` with a slimstat-local sanitizer that strips only control chars and tags. The existing `parse_url` + host regex + 2048-char cap already guard against injection for the in-scope schemes.
2. **Better (idiomatic WP)** — add a `kses_allowed_protocols` filter callback in the tracker bootstrap that appends `android-app` (and `ios-app` for iOS deep-link referers) only for the duration of the tracking request. Keeps `sanitize_url()` in place, satisfies WPCS, and naturally covers any future scheme expansion.
3. **Defensive add-on** — store the original raw value as a fallback in `$stat['notes']` when `sanitize_url()` returns empty for a non-empty input, so analytics aren't silently zeroed even if a future scheme slips through.

A regression test belongs alongside the fix:

- E2E: simulate an Android-app referer via Playwright `page.setExtraHTTPHeaders({ Referer: 'android-app://com.google.android.googlequicksearchbox/' })`, trigger a pageview, assert the DB `referer` column.
- PHP integration: post a base64URL-encoded `android-app://...` to the tracker endpoint, assert stored row.

## Severity

**HIGH** — Google Discover is a major mobile traffic source for many content sites. Silent data loss since the v5.4.0 release (Oct 2025). All Android-app referers (Google app, Twitter, Reddit, Facebook, etc.) are affected.

## Source

Support thread: wp.org forum, reporter `dxylott54`, language: Japanese (with English translation provided). The reporter also documented Issue 1 (search box) which is a duplicate of #298.
